cpp/oneapi/dal/backend/primitives/rng/device_engine_dpc.cpp

```diff
@@ -105,17 +131,53 @@ sycl::event uniform_without_replacement(sycl::queue& queue,
                                         Type a,
                                         Type b,
                                         const event_vector& deps) {
-    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) ==
-        sycl::usm::alloc::device) {
+    if (sycl::get_pointer_type(dst, engine_.get_queue().get_context()) == sycl::usm::alloc::host) {
         throw domain_error(dal::detail::error_messages::unsupported_data_type());
     }
-    void* state = engine_.get_host_engine_state();
-    engine_.skip_ahead_gpu(count);
-    uniform_dispatcher::uniform_without_replacement_by_cpu<Type>(count, dst, state, a, b);
-    auto event = queue.submit([&](sycl::handler& h) {
-        h.depends_on(deps);
+    const std::int64_t n = b - a;
+    if (count > n) {
+        throw std::invalid_argument("Sample size exceeds population size");
+    }
+
+    // Allocate array for full population [a, b)
+    auto full_array = array<Type>::empty(queue, n);
+    Type* full_ptr = full_array.get_mutable_data();
+
+    // Fill array with [a, a+1, ..., b-1]
+    auto fill_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(sycl::range<1>(n), [=](sycl::id<1> i) {
+            full_ptr[i] = a + static_cast<Type>(i);
+        });
     });
-    return event;
+
+    // Generate random indices for Fisher-Yates shuffle
+    auto rand_array = array<Type>::empty(queue, count);
+    Type* rand_ptr = rand_array.get_mutable_data();
+
+    oneapi::mkl::rng::uniform<Type> distr(0, n);
```

Comment:
Please mention somewhere in the description that this implementation only works with `std::int32_t` and `std::uint32_t` as `Type`, because oneMKL supports only those 2 types for discrete uniform distribution.

Or use `oneapi::mkl::rng::uniform<std::int32_t>` here like it is done in `partial_fisher_yates_shuffle`. In this case less restrictions on Type are required.